lib/puma/cluster/worker.rb

```diff
@@ -57,21 +57,26 @@ def run
         @config.run_hooks(:before_worker_boot, index, @log_writer, @hook_data)
 
         begin
-        server = @server ||= start_server
+          server = @server ||= start_server
         rescue Exception => e
           log "! Unable to start worker"
           log e
           log e.backtrace.join("\n    ")
           exit 1
         end
 
-        restart_server = Queue.new << true << false
+        restart_server = Queue.new << Puma::Const::WorkerCmd::RESTART << Puma::Const::WorkerCmd::STOPPED
 
         fork_worker = @options[:fork_worker] && index == 0
 
+        # worker ids for validte hang process
+        worker_pids = []
+
+        # create a mutex to synchronize the server thread
+        mutex = Mutex.new
```

Comment:
I'm not clear on what this buys us; nothing about the new state in the queue seems to require synchronizing around this and it messes up the signal trap hooks pretty badly. Which race condition were you worried about here?

Reply:
There's probably a narrower fix for CI, given the unlikelihood of the race condition in the real world; most setups are likely going to send a second signal if it doesn't terminate in time, and nobody restarts worker zero as frequently or quickly as tests do (ideally). As mentioned above if we think we need the mutex, we should change the restart_server queue to a simpler state machine, but I think the easier fix is to continue to avoid the mutex and add a second TERM to the specs.